btb/recommendation/matrix_factorization.py

```diff
@@ -0,0 +1,117 @@
+import logging
+
+import numpy as np
+import scipy.stats as stats
+from sklearn.decomposition import NMF
+
+from btb.recommendation.recommender import BaseRecommender
+
+LOGGER = logging.getLogger('btb')
+
+
+class MFRecommender(BaseRecommender):
+    """
+    Recommender for recomending pipelines to try on a new dataset D based
+    on performances of datasets on the different pipeline options. Recommends
+    pipelines that would maximize the score value. Uses Matrix Factorization
+    to determine which pipeline to recommend.
+
+    Attributes:
+        dpp_matrix: (2D np.array) Dataset performance pipeline matrix. Num
+            datasets by num pipelines matrix where each row i corresponds to a
+            dataset and each collumn j corresponds to a pipeline. The
+            dpp_marix[i,j] is the score of pipeline j on dataset i or 0 if
+            pipeline j has not beentried on dataset i.
+        n_components: (int) The number of components (columsn) to keep after
+            Matrix Factorxation.
+        mf_model: Matrix Factorization model that reduces dimensionailty from
+            num pipelines space to n_components space.
+        dpp_ranked: (2D np.array) Matrix of rankings for each row of dpp_matrix
+            after matrix facorization has been applied.
+        dpp_vector: (1D np.array) Vector representing pipeline performances on
+            a new dataset D.
+        matching_dataset: (1D np.array) Row from dpp_matrix representing
+            pipeline performances for the dataset that most closely matches the
+            new dataset D. Identified in fit.
+    """
+
+    def __init__(self, dpp_matrix, n_components=100, **kwargs):
+        """
+        Args:
+            dpp_matrix: np.array shape = (num_datasets, num_pipelines) Sparse
+                dataset performance matrix pertaining to pipeline
+                scores on different dataset. Each row i coresponds to a dataset
+                and each column j corresponds to a pipeline. dpp_matrix[i,j]
+                corresponds to the score of the pipeline j on the dataset and
+                is 0 if the pipeline was not tried on the dataset
+            n_components: int. Corresponds to the number of features to keep
+                in matrix decomposition. Must be >= number of rows in matrix
+        """
+        self.dpp_matrix = dpp_matrix
+        self.n_components = n_components
+        self.mf_model = NMF(
```

Comment:
This is just a personal styling preference: When line length start approaching the 99 chars, switching to multi-line method calling (either one or multiple argument per line) is OK.
But when the line stays short, and there are only two or three arguments, keeping single-line method calling makes code shorter and quicker to read.